mm2src/coins/tendermint/tendermint_coin.rs

```diff
@@ -237,7 +237,7 @@ pub struct TendermintCoinImpl {
     pub(crate) history_sync_state: Mutex<HistorySyncState>,
     client: TendermintRpcClient,
     chain_registry_name: Option<String>,
-    pub(crate) ctx: MmArc,
+    pub(crate) ctx: MmWeak,
```

Comment:
Mm context is the central point of channel controller, so it can be accessed at anytime on runtime without requiring anything extra. If we clone that into coin structs, this will lead another self-referencing pointer https://github.com/KomodoPlatform/komodo-defi-framework/pull/1978#discussion_r1348240562

Reply:
> If we clone that into coin structs, this will lead another self-referencing pointer

I don't think since the ref counter will be decreased when the coin is deactivated. Unlike the context which has the coin context as part of it.

Reply:
Oh true, it doesn't cause that. But I am still not sure if it's a good idea. We will basically need to add this controller to all coin types if we follow that approach.

Currently all we do is upgrading the weak pointer(which is one-time op) on the coin activation and thats it.

Reply:
> Oh true, it doesn't cause that. But I am still not sure if it's a good idea. We will basically need to add this controller to all coin types if we follow that approach.

But we will add `MmWeak` to all coins otherwise.

> Currently all we do is upgrading the weak pointer(which is one-time op) on the coin activation and thats it.

I understand and I don't mind that, my concern is that having `MmWeak` in coins will cause devs to use it to access other fields from `MmArc` in the future, this will lead to unneeded complexities.

Reply:
> But we will add MmWeak to all coins otherwise.

Eth already have it included, and it's much useful(provides the complete context access) than the event controller.

> I understand and I don't mind that, my concern is that having MmWeak in coins will cause devs to use it to access other fields from MmArc in the future, this will lead to unneeded complexities.

I agree about the complexity issue, but not having access to the app context is not a solution. We shouldn't include these types of fields in the coin context. They are more related to the app itself rather than the coin protocol. We need to work on the context and coin protocol types to address this in a proper way(e.g., being able to access mm context directly from anywhere in the application).

Reply:
> They are more related to the app itself rather than the coin protocol

I agree with you on this regarding `stream_channel_controller`

> We need to work on the context and coin protocol types to address this in a proper way(e.g., being able to access mm context directly from anywhere in the application).

How it has been handled in other coins was that we just pass what we need as parameters and decouple global context from coins, `Eth` was an old implementation that's why it was done differently. An example for this is how we pass sqlite connection from context to a lot of places, sqlite connection is also related to the whole app and not a single coin protocol. My idea was to rely on a single approach until we can find a better way to access mm context directly from anywhere in the application as you suggested. I won't fight this anymore though :), we can keep `MmWeak` in tendermint coin for now and think about this for future implementations.

Reply:
Sure, I don't have hard feelings on this topic. I can replace the weak pointer right away if it's a critical concern.